src/iblt.h

```diff
@@ -130,16 +161,52 @@ class CIblt
             READWRITE(salt);
         }
 
-        if (ser_action.ForRead() && version > 1)
-            throw std::ios_base::failure("No IBLT version exceeding 1 is currently known.");
+        if (ser_action.ForRead() && version > IBLT_MAX_VERSION_SUPPORTED)
+            throw std::ios_base::failure("No IBLT version exceeding 2 is currently known.");
 
         READWRITE(n_hash);
         if (ser_action.ForRead() && n_hash == 0)
         {
             throw std::ios_base::failure("Number of IBLT hash functions needs to be > 0");
         }
         READWRITE(is_modified);
-        READWRITE(hashTable);
+
+        if (version >= 2)
+        {
+            READWRITE(keycheckMask);
+            READWRITE(hashTable);
+        }
+        else
+        {
+            std::vector<HashTableEntryStaticChk> hashTableChk;
+            if (ser_action.ForRead())
+            {
```

Comment:
There is nothing preventing deserialization of checksums that are outside the range of the `keycheckMask` from the network.

As far as I can see, this has no security implications yet whatsoever but maybe it is a good idea to always strip off extra bits when checksums get deserialized from the network, to ensure that they also conform to what is internally expected by the IBLT.

With functions to stuff/unstuff the individual odd-sized checksum bits from a byte array as I suggested above, this functionality could of course be outsourced into that deserialization function.

Overall I also wonder whether the class hierarchy is necessary here; or do you plan on further extending the types of hash tables? To be honest it all looks to me like it would be more clear and concise in a single class still.

Reply:
So are you suggesting that, during read and for version >= 2, I should similarly cycle through hash table entries and use the update: `entry.keyCheck = entryChk.keyCheck & keycheckMask`? That would definitely be possible when deserializing CIblt. But when you say outsourced into "that" deserialization function, do you mean the HashTableEntry deserialization function? That is not currently possible because HashTableEntry does not keep track of keycheckMask. Of course this could be added, but at the cost of carrying around an extra 32 bits of data per entry (only locally of course).

Originally I intended to maintain just a single HashTableEntry class. However, because this class has no native version field, I ran into trouble instantiating the proper version for each instance when serializing / deserializing from / to the legacy implementation. For example, deserializing vector<HashTableEntry> in CIblt from a sender running old code requires that the deserialization code in HashTableEntry knows to use version 0. But there is no way to communicate what version to use because the instantiation vector<HashTableEntry> uses the default constructor, which accepts no arguments. The only way to get around this, that I could think of, was to instead manually allocate the vector with dynamically deserialized HashTableEntry objects. First I would create the HashTableEntry object having proper version, then I would call READWRITE on that, at which point the object's deserialization procedure would be aware of the version, and then push the object into the vector. This seemed at least as ugly to me, although opinions may vary :-).

Reply:
I went ahead and made the `entry.keyCheck = entry.keyCheck & keycheckMask` change I suggested above.